### ISSUE TITLE
Version selection logic fails with small/old version numbers

### ISSUE THREAD

**Issue by DeeDeeG** [tj/n#570]
### Describe issue including what OS you are using

Can't install old versions (0.7 or older)

(Using Ubuntu 19.10...)

### Describe what version of N you have

4.1.0, but also happens on develop branch (5.0.0-next.0)

### Describe how you installed N

I cloned https://github.com/tj/n, and put a symlink to the n binary in my /usr/local/bin folder:
`sudo ln -s /home/[me]/n/bin/n /usr/local/bin/n`.

This way I can switch n versions by checking out different git branches in my local n repo.

I also made sure n versions are installed within my home directory:
`export N_PREFIX=/home/[me]/n-prefix`

### Steps to reproduce issue

1. `n 0.0.1` _(also applies when running `n 0.0.2`, all the way through `n 0.0.6`)_
2. Error: invalid version '0.0.1'

---

1. `n 0.1`
2. _node version 0.12.18 is installed instead_

---

1. `n 0.1.0` _(also applies when running `n 0.1.1`, all the way through `n 0.1.104`)_
2. Error: invalid version '0.1.1'

---

1. `n 0.2` _(also applies when running `n 0.3`, all the way through `n 0.7`)_
2. Error: invalid version '0.2'

---

1. `n 0.8.1`
2. _node version 0.8.19 is installed instead_

--

1. `n 0.8.2`
2. _node version 0.8.28 is installed instead_

--

1. `n 0.8.3` _(also applies when running `n 0.8.4`, or `n 0.8.5`)_
2. Error: invalid version '0.8.3'

--

1. `n 1`
2. _node version 12.4.0 is installed instead_

### Describe the results you received

Some older versions are considered invalid... even though it doesn't say these old versions aren't supported by n.

And in some cases, small numbers are interpreted/expanded unexpectedly: "1" becomes "12" or "19", 2 becomes "28".

### Describe the results you expected

Old versions should download and install without problems, OR they should be expressly not supported (ideally with documentation to go along with it, maybe just a mention in the README).

Weird corner case with `n 0.8.3` through `n 0.8.5` should be handled.

Also, small numbers like "1" or "2" shouldn't be interpreted/expanded as "12" "19" or "28".

### What version of npm you are using

I'm not invoking/using npm here

### Additional information you deem important (e.g. issue happens only occasionally):

My active version of node always comes up as 10.16.0, which I have already installed via nvm... I don't think that changes any of the issues described, however.

**Comment by shadowspawn** [comment]
There are multiple things going on. I don't expect many people are trying those old versions, so not planning to do work to extend support for them, but some of the things you saw may apply to current versions too. 

1) n is downloading tarballs with the modern folder layout from `https://nodejs.org/dist/`. Some of the early versions you are trying only have a single tarball at the top level and not a folder, and are not supported. e.g. `0.10.0`

2) The error message "invalid version" is currently a bit of a catch-all for multiple problems (and I intend to improve that but requires some refactoring).

3) The general intent is that `n 4` is treated as select newest version of `4.x`, and `n x.y.z` is taken as being explicit. I don't have an immediate explanation for the odd number changes you are seeing, and want to dig into that further.

(Thanks for trying on develop branch!)

**Comment by DeeDeeG** [comment]
> There are multiple things going on.

Yeah, that must be true.

> I don't expect many people are trying those old versions, so not planning to do work to extend support for them,

Makes sense to me. I just felt a sort of twinge of "but the program should be correct!" followed by "using node older than v1.0 in 2019 is kind of a horrifying thought." Decided to file an issue because "why not."

> but some of the things you saw may apply to current versions too.

I can confirm: e.g. running `n 10.1` installs node 10.16.0 instead.

> 1. n is downloading tarballs with the modern folder layout from https://nodejs.org/dist/. Some of the early versions you are trying only have a single tarball at the top level and not a folder, and are not supported. e.g. `0.10.0`

I believe these versions all have folders with tarballs at the moment. I see the tarballs at nodejs.org/dist/[tarball], but also they have subfolders, e.g. nodejs.org/dist/v0.10.0/ . Maybe this is a recent development. I am pretty new to working with node.

> 2. The error message "invalid version" is currently a bit of a catch-all for multiple problems (and I intend to improve that but requires some refactoring).

I noticed this when working on my PR. I see the exact same text in two different error handling places in the code. I hope they are changed to have unique text, as a first step. In my PR, I was pretty sure the second instance of this error message was un-triggerable.

I have a hunch that these block of logic could be re-implemented/checked for correctness: https://github.com/tj/n/blob/develop/bin/n#L680-L688, https://github.com/tj/n/blob/develop/bin/n#L528-L537, and the display_latest_version() and similar. Probably involves reading some arcane bash documentation... and double-checking the exact behavior in `man grep`. If I can figure this out, maybe I can be of some help in fixing it.

**Comment by shadowspawn** [comment]
Oops, I got wrong version as example

>1. n is downloading tarballs with the modern folder layout from https://nodejs.org/dist/. Some of the early versions you are trying only have a single tarball at the top level and not a folder, and are not supported. e.g. 0.10.0

The example should have been 0.1.0 which does not have a folder.

**Comment by DeeDeeG** [comment]
> The example should have been 0.1.0 which does not have a folder.

Fair enough. What is strange to me is that 0.1.14 (https://nodejs.org/dist/v0.1.14/) through 0.1.33, and 0.1.90 through 0.1.104 do have folders, yet n is not downloading them. 🤷

If I may point out, `n ls` gives all versions back to 0.0.1. I feel as if this goes against expectations; I tend to presume these versions are all installable with n if the installer tool lists all the versions. Logic for `n ls` could potentially be updated, so as to narrow it down to supported versions. Again, if I can figure this out and implement it well, maybe I can contribute a PR to do so.

**Comment by shadowspawn** [comment]
> I have a hunch that these block of logic could be re-implemented/checked for correctness: https://github.com/tj/n/blob/develop/bin/n#L680-L688, https://github.com/tj/n/blob/develop/bin/n#L528-L537, and the display_latest_version() and similar. 

Yes, fun with regular expressions. I made multiple iterations of improvements in this area in my fork [nvh](https://github.com/shadowspawn/nvh), but ultimately switched to downloading `index.tab` instead of screen scraping the html pages which changed the version lookup entirely. I intend to do this for `n` too. 

I agree with you that `n ls` should be consistent with the installable versions. (And less noisy by default: #383) Changing to `index.tab` also makes restricting version lookup to supported versions for current platform much more straightforward.

**Comment by DeeDeeG** [comment]
(at risk of being slightly off-topic, I noticed your fork, nvh, and I like the code there! A bit more readable, probably easier to work with/verify its correctness etc. Maybe it should be the future of n, and n can be phased out? Or if n should stick around to give continuity for existing users, perhaps some code changes can be back-ported from nvh to n, just without changing the existing command-line commands and their end results (beyond bugfixes/minor quality of life improvements].)

**Comment by shadowspawn** [comment]
(Thanks! My current plan is selective back-porting from `nvh` to `n`. `n` has much higher installed base and visibility due to historical references, including mentions in npmjs documentation.)

**Comment by shadowspawn** [comment]
I dug into the partial version number matching and it is indeed not matching on "whole" numbers and prone to finding later versions, as you found with "10.1" matching "10.16.0".

I tracked down a relevant changeset from nvh: https://github.com/shadowspawn/nvh/commit/91cabee5103d6d74eef82826448bc2626906b836#diff-c1111bd512b29e821b120b86446026b8

The partial matches in n are currently just looking for version starting with string, rather than matching the "whole" number by including a trailing dot.

c.f. `egrep ^$version` vs `egrep "^$partial\."

**Comment by shadowspawn** [comment]
So many little issues... The version lookup is triggering when it should not. This looks correct, but does not work as intended.

```
local dots="${version//[^.]*/}"
```

removing the * fixes it so collapses to just dots: 

```
local dots="${version//[^.]/}"
```

but the versions before 0.8.6 do not have binaries according to #95 #346 #350

(Enough digging for today, made sense of most of the oddities!)

**Comment by DeeDeeG** [comment]
So to summarize: 
- n can't operate on versions below 0.8.6.
  - Not a bug (old versions are hard to support, probably not worth the effort and script complexity to attempt compiling them from source.)
  - Could be clearer in documentation, at least the README
  - Specific error message for these versions would be great IMO (for example "versions less than 0.8.6 are not supported by n")
- single digits getting expanded to double digit numbers is actually a bug

**Comment by DeeDeeG** [comment]
> This looks correct, but does not work as intended.
> 
> `local dots="${version//[^.]*/}"`
> 
> removing the * fixes it so collapses to just dots:
> 
> `local dots="${version//[^.]/}"`

In my testing, the first variant always has `$dots` become an empty string. Thus the version filtering block below will ALWAYS run, and will effectively sanitize/validate `$version`.

We should be careful if fixing the dots logic, since it can allow all kinds of un-sanitized values of `version` from the user input into later parts of the script. Perhaps the "dots" check can be removed entirely. That way all user input for `$version` is sanitized. The only downside is some milliseconds* of doing grep and running regex against a string. 

\* (Unless someone inputs a massive string as the version... e.g. `n $(cat war_and_peace_full_novel_no_spaces_or_newlines.txt)` We can prevent this by truncating the user input to the first _n_ characters, maybe 10 or 20 or less.)

**Comment by shadowspawn** [comment]
Released v5.0.0 with fixes originating from this issue.

Thank you  for your contributions.
